python/dgl/node_flow.py

```diff
@@ -0,0 +1,845 @@
+            a :mod:`Node UDF <dgl.udf>`.
+        block_id : int or ALL
+            the block to register the apply node function.
+        """
+        if is_all(block_id):
+            self._apply_node_funcs = [func] * self.num_blocks
+        else:
+            self._apply_node_funcs[block_id] = func
+
+    def register_apply_edge_func(self, func, block_id=ALL):
+        """Register global edge apply function for a block.
+
+        Once registered, ``func`` will be used as the default apply
+        edge function in :func:`apply_block`.
+
+        Parameters
+        ----------
+        func : callable
+            Apply function on the edge. The function should be
+            an :mod:`Edge UDF <dgl.udf>`.
+        block_id : int or ALL
+            the block to register the apply edge function.
+        """
+        if is_all(block_id):
+            self._apply_edge_funcs = [func] * self.num_blocks
+        else:
+            self._apply_edge_funcs[block_id] = func
+
+    def apply_layer(self, layer_id, func="default", v=ALL, inplace=False):
+        """Apply node update function on the node embeddings in the specified layer.
+
+        Parameters
+        ----------
+        layer_id : int
+            The specified layer to update node embeddings.
+        func : callable or None, optional
+            Apply function on the nodes. The function should be
+            a :mod:`Node UDF <dgl.udf>`.
+        v : a list of vertex Ids or ALL.
+            The vertices to run the node update function.
+        inplace : bool, optional
+            If True, update will be done in place, but autograd will break.
+        """
+        if func == "default":
+            func = self._apply_node_funcs[block_id]
+        if is_all(v):
+            v = utils.toindex(slice(0, self.layer_size(layer_id)))
+        else:
+            v = v - int(self._layer_offsets[layer_id])
+            v = utils.toindex(v)
+        with ir.prog() as prog:
+            scheduler.schedule_nodeflow_apply_nodes(graph=self,
+                                                    layer_id=layer_id,
+                                                    v=v,
+                                                    apply_func=func,
+                                                    inplace=inplace)
+            Runtime.run(prog)
+
+    def _layer_local_nid(self, layer_id):
+        return F.arange(0, self.layer_size(layer_id))
+
+    def apply_block(self, block_id, func="default", edges=ALL, inplace=False):
+        """Apply edge update function on the edge embeddings in the specified layer.
+
+        Parameters
+        ----------
+        block_id : int
+            The specified block to update edge embeddings.
+        func : callable or None, optional
+            Apply function on the nodes. The function should be
+            a :mod:`Node UDF <dgl.udf>`.
+        e : a list of edge Ids or ALL.
+            The edges to run the edge update function.
```

Comment:
`edges` instead of `e`; Is the id local or global? Does this support (u, v) tuple? If (u, v) tuple, local or global?